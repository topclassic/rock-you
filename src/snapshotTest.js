
import renderer from 'react-test-renderer'

const snapshotTest = (testCase, element) => {
  it(testCase, () => {
    const wrapper = renderer.create(element).toJSON()
    expect(wrapper).toMatchSnapshot()
  })
}

export default snapshotTest