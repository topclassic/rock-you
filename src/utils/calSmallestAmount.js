import { bill, cent } from "../constants/amount";
import pluralize from "pluralize";

const convertDigit = (value, digit) => {
  return `${value
    .toString()
    .split("")
    .reverse()
    .join("")
    .charAt(digit)}`;
};

const checkPluralNoun = (name, value) => {
  if (value <= 1) {
    return name;
  }
  return pluralize.plural(name);
};

const checkBill = value => {
  let txt = "";
  let data = "";
  let calDigi = 0;
  const arrObj = Object.keys(bill);
  let digitTen = parseInt(
    `${convertDigit(value, 1)}${convertDigit(value, 0)}`,
    10
  );
  arrObj.forEach((obj, index) => {
    if (index === 0) {
      if (value >= bill[obj]) {
        calDigi = `${parseInt(value / bill[obj], 10)}`;
        data = `${calDigi} ${bill[obj]} dollar ${checkPluralNoun(
          "bill",
          calDigi
        )}`;
        txt = txt ? `${txt}, ${data}` : `${data}`;
      }
    } else {
      if (digitTen >= bill[obj]) {
        calDigi = parseInt(digitTen / bill[obj], 10);
        data = `${calDigi} ${bill[obj]} dollar ${checkPluralNoun(
          "bill",
          calDigi
        )}`;
        txt = txt ? `${txt}, ${data}` : `${data}`;
        digitTen = digitTen - bill[obj] * calDigi;
      }
    }
  });
  return txt;
};

const checkCent = value => {
  let txt = "";
  let data = "";
  let calDigi = 0;
  let digi = value;
  const arrObj = Object.keys(cent);
  arrObj.forEach(obj => {
    if (digi >= cent[obj]) {
      calDigi = parseInt(digi / cent[obj], 10);
      data = `${calDigi} ${checkPluralNoun(obj, calDigi)}`;
      txt = txt ? `${txt}, ${data}` : `${data}`;
      digi = digi - cent[obj] * calDigi;
    }
  });
  return txt;
};

const calculate = value => {
  const arrValue = value.split(".");
  let amount;
  if (arrValue.length > 1) {
    const digit = checkBill(parseInt(arrValue[0], 10));
    const decimal = checkCent(parseInt(arrValue[1], 10));
    amount = digit
      ? `Your change is ${digit}, ${decimal}`
      : `Your change is ${decimal}`;
  } else {
    amount = `Your change is ${checkBill(parseInt(arrValue[0], 10))}`;
  }
  const arrAmount = amount.split(",");
  arrAmount.splice(
    arrAmount.length - 1,
    1,
    ` and${arrAmount[arrAmount.length - 1]}`
  );
  return arrAmount.toString();
};

export default calculate;
