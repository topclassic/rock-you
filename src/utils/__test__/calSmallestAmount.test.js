import calSmallestAmount from "../calSmallestAmount";
import snapshotTest from "../../snapshotTest";

describe("CalSmallestAmount", () => {
  it("test func calculate amount correctly", () => {
    expect(calSmallestAmount("100")).toEqual(
      "Your change is 1 100 dollar bill"
    );
    expect(calSmallestAmount(".99")).toEqual(
      "Your change is 3 quarters, 2 dimes, and 4 pennies"
    );
    expect(calSmallestAmount("124.67")).toEqual(
      "Your change is 1 100 dollar bill, 1 20 dollar bill, 4 1 dollar bills, 2 quarters, 1 dime, 1 nickel, and 2 pennies"
    );
  });
});

describe("CalSmallestAmount testing", () => {
  snapshotTest("func CalSmallestAmount(100)", calSmallestAmount("100"));
  snapshotTest("func CalSmallestAmount(.99)", calSmallestAmount(".99"));
  snapshotTest("func CalSmallestAmount(124.67)", calSmallestAmount("124.67"));
});
