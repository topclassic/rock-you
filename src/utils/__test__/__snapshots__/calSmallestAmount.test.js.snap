// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalSmallestAmount testing func CalSmallestAmount(.99) 1`] = `"Your change is 3 quarters, 2 dimes, and 4 pennies"`;

exports[`CalSmallestAmount testing func CalSmallestAmount(100) 1`] = `"Your change is 1 100 dollar bill"`;

exports[`CalSmallestAmount testing func CalSmallestAmount(124.67) 1`] = `"Your change is 1 100 dollar bill, 1 20 dollar bill, 4 1 dollar bills, 2 quarters, 1 dime, 1 nickel, and 2 pennies"`;
